cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,447 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from dataclasses import dataclass
+from typing import Optional, Sequence, Union, Tuple, List, Type, cast, TYPE_CHECKING, Any
+
+import itertools
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@dataclass(frozen=True, eq=False)  # type: ignore
+class ControlValues(abc.ABC):
+    """ControlValues is an abstract data class that represents different control values."""
+
+    _combinations: Any
+    _nxt: Optional['ControlValues'] = None
```

Comment:
this assumes that lists of combinations are the only way that wil be used to define control_values, there was a previous comment on this PR about possibly supporting other ways like tensors or sympy expressions in the future.

the linked-list structure allows defining using different ways at the same time e.g. CV1(a table) -> CV2(a sympy expression) -> ... etc

each of those object will inheret from the ControlValues base class so that wil work perfectly.

Reply:
The comment referred to above is on the linked issue. https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817

Reply:
Another thought, have you considered making a separate `ControlValuesProduct` class that merges two (or maybe N) ControlValues objects into a product? That approach would allow for a smaller base class and adhere more to the single responsibility pattern.